doc/design.md

```diff
@@ -0,0 +1,949 @@
+For small data the threading overhead may be larger than the gains.
+For large data, many cases such as `operator+` are heavily limited by memory bandwidth and using more cores than necessary for reaching the peak bandwidth would be wasteful.
+
+
+### <a name="design-details-no-magic"></a>No magic
+
+##### Rationale
+
+Many algorithms in Mantid do "magic" under the hood:
+- Convert inputs if required. *Example: Conversion from point data in `Algorithms::ConvertUnits` and `Algorithms::Rebin`.*
+- Tolerate partially bad input data. *Example: Replacing special values in `Algorithms::SumSpectra`.*
+- Support doing two different things that are actually different. *Example: `Algorithms::Rebin` can rebin histogram data, bin event data, or set new bin edges in a workspace.*
+- Combine multiple processing steps. *Examples: Aligning bins in `Algorithms::ConvertUnits`. Loading and saving calibration files in `Algorithms::MaskDetectorsIf`.*
+
+Support for such features complicates the implementation and makes reasoning about code harder.
+It should therefore not be done in this library.
+- Operations should simply fail as early as possible.
+- Inputs are required to be cleaned up and converted to the correct format explicitly.
+- Processing steps should be kept as separate operations.
+
+##### Note
+
+There is nothing wrong with "magic" in higher-level operations.
+High level algorithms that simply work no matter what is thrown at them are likely still a requirement.
+We are merely banning it from *this* library, i.e., the lowest-level building block.
+
+
+### <a name="design-details-no-logging"></a>No logging
+
+##### Rationale
+
+As a corollary from the "no magic" strategy above, there will be very few occasions that require log messages.
+Logging can be implemented in higher-level algorithms, or potentially in some Python wrappers.
+Adding logging functionality would increase complexity and add extra dependencies.
+
+##### Workaround
+
+If required we could log to `stdout` using `std::cout`, which can be redirected and filtered.
+
+
+### <a name="design-details-exception-safety"></a>Operations have strong exception guarantee or clear data if only a basic exception guarantee can be given
+
+##### Rationale
+
+For performance reasons operations can be (and are in practice) applied in-place.
+If an exception is thrown it can happen that data is partially modified and thus in an unknown state.
+It is impossible for client code to tell if data has been corrupted, unless there is a well defined exception safety guarantee in combination with flagging of bad data.
+The user thus continue working with the bad data, with unknown consequences.
+
+##### Implementation
+
+Wherever possible, operations should provide a strong exception guarantee, i.e., even for in-place operations there should be no risk of unknowingly corrupting data.
+This simply boils down to performing any checks up-front, before data is modified.
```

Comment:
Yes, should perform checks up front, but that's not enough to label it a strong guarantee.

Reply:
Is it not?
"Strong exception guarantee -- If the function throws an exception, the state of the program is rolled back to the state just before the function call." (https://en.cppreference.com/w/cpp/language/exceptions).

I check up-front, so I can throw before modification, i.e., no rollback is required. Once I start modifying I know there cannot be further exceptions, so I do not need to implement rollback.